src/Polysemy/Compose.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+module Polysemy.Compose
+  ( Compose (..)
+  , sendCompose
+  , runCompose
+  , subsumeCompose
+  , KnownList
+  ) where
+
+import Data.Proxy
+import Polysemy
+import Polysemy.Internal
+import Polysemy.Internal.Union
+
+data Compose :: EffectRow -> Effect where
+  Compose :: ElemOf r e -> e m a -> Compose r m a
+
+sendCompose
+  :: forall e r' r a
+   . (Member e r', Member (Compose r') r)
+  => Sem (e ': r) a
+  -> Sem r a
+sendCompose = hoistSem $ \u -> case decomp u of
+  Right (Weaving e s wv ex ins) ->
+    injWeaving $
+      Weaving (Compose (membership @e @r') e) s (sendCompose @e @r' . wv) ex ins
+  Left g -> hoist (sendCompose @e @r') g
+
+type family Append l r where
+  Append (a ': l) r = a ': (Append l r)
+  Append '[] r = r
+
+runCompose
+  :: forall r' r a
+   . KnownList r'
+  => Sem (Compose r' ': r) a
+  -> Sem (Append r' r) a
+runCompose = hoistSem $ \u -> hoist runCompose $ case decomp u of
+  Right (Weaving (Compose pr e) s wv ex ins) ->
+    Union (extendMembership @_ @r pr) $ Weaving e s wv ex ins
+  Left g -> weakenList @r' @r g
+
+
+subsumeMembership :: forall r r' e. Members r r' => ElemOf r e -> ElemOf r' e
+subsumeMembership Here = membership @e @r'
+subsumeMembership (In (pr :: ElemOf r'' e)) = subsumeMembership @r'' @r' pr
+
+subsumeCompose
+  :: forall r' r a
+   . Members r' r
+  => Sem (Compose r' ': r) a
+  -> Sem r a
+subsumeCompose = hoistSem $ \u -> hoist subsumeCompose $ case decomp u of
+  Right (Weaving (Compose pr e) s wv ex ins) ->
+    Union (subsumeMembership pr) (Weaving e s wv ex ins)
+  Left g -> g
+
+extendMembership :: forall r r' e. ElemOf r e -> ElemOf (Append r r') e
+extendMembership Here = Here
+extendMembership (In e) = In (extendMembership @_ @r' e)
+
+weakenList :: forall r' r m a
+            . KnownList r'
+           => Union r m a
+           -> Union (Append r' r) m a
+weakenList u = unconsKnownList @_ @r' u (\_ (_ :: Proxy r'') -> weaken (weakenList @r'' u))
+
+class KnownList (l :: [k]) where
```

Comment:
Do you have some source/writeup/blogpost for this - I haven't encountered it before it and it seems rather interesting?(especially the constraints in `unconsKnownList`)

Reply:
I adapted a trick I learned ways back (don't remember where from) for pattern-matching on type-level expressions at run-time.
I don't actually know of a writeup for this, but the original pattern came from the following way of encoding type-level naturals:

```haskell
data Nat = Z | S Nat

class KnownNat (n :: Nat) where
  induceNat :: (n ~ 'Z => a) -> ((n ~ ('S n'), KnownNat n') => Proxy n' -> a) -> a

instance KnownNat 'Z where
  induceNat b _ = b

instance KnownNat n => KnownNat ('S n) where
  induceNat _ i = i Proxy
```

Basically, given a type-level `Nat`, you can do induction on it by providing what you'll do in case it's `'Z` -- and when defining that case, you get to know that `n ~ 'Z` -- and what you'll do in case it's `'S n'` -- and in that case, you get to know that `n ~ 'S n'`. This is the same pattern, except for lists instead.

I think nowadays this pattern has been replaced by [singletons](https://hackage.haskell.org/package/singletons), although I'm not sure.

Reply:
It seems to be this symmetry between existential wrapper/CPS - singleton basically encodes those possible cases as datatype instead of call to continuation.